### ISSUE TITLE
Any plan to release the training code?

### ISSUE THREAD

**Issue by hmyang0727** [ByteDance-Seed/Depth-Anything-3#112]
Hi, thank you for your great work!

I’m wondering if there is any plan or timeline for releasing the training code along with dataloaders. It would be extremely helpful for reproducing results and experimenting further.

Thanks in advance!

**Comment by Golbstein** [comment]
I've written a dataloader for preprocessed waymo dataset:

```python
import glob
import os
import pickle
import numpy as np
import scipy.sparse
import torch
from PIL import Image
from torch.utils.data import Dataset
from torchvision import transforms

from .dataset_utils import process_rgb_and_depth

# Coordinate System Adapter
# DA3/OpenCV expects rays to start in Optical Frame: X=Right, Y=Down, Z=Forward
# Waymo Camera Frame is: X=Forward, Y=Left, Z=Up
# We need a matrix to rotate Optical -> Waymo Camera
# X_waymo = Z_opt
# Y_waymo = -X_opt
# Z_waymo = -Y_opt

T_waymo2colmap = np.array([[0, 0, 1, 0], [-1, 0, 0, 0], [0, -1, 0, 0], [0, 0, 0, 1]])


class DrivingSegmentDataset(Dataset):
    def __init__(self, root_dir, camera_setup='all', seq_len=10, image_size=(336, 504), original_fps=10, target_fps=10, debug: bool = True):
        """
        Args:
            original_fps (int): FPS of the raw data (default ~10).
            target_fps (int): Desired FPS for the network input.
        """
        self.debug = debug
        self.root_dir = root_dir
        self.seq_len = seq_len
        self.target_size = image_size

        # --- FPS Logic ---
        if target_fps > original_fps:
            raise ValueError(f"Target FPS ({target_fps}) cannot be higher than Original FPS ({original_fps})")

        self.original_fps = original_fps
        self.target_fps = target_fps
        self.step = max(1, int(original_fps / target_fps))  # Stride

        # 1. Find all segment folders
        self.segments = sorted(glob.glob(os.path.join(root_dir, 'segment-*')))
        if camera_setup == 'all':
            self.camera_names = ['FRONT', 'FRONT_LEFT', 'FRONT_RIGHT', 'SIDE_LEFT', 'SIDE_RIGHT']
        elif camera_setup == 'front':
            self.camera_names = ['FRONT', 'FRONT_LEFT', 'FRONT_RIGHT']
        elif camera_setup == 'front_only':
            self.camera_names = ['FRONT']

        self.transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        ])

        self.h, self.w = image_size
        y, x = torch.meshgrid(
            torch.arange(self.h / 2, dtype=torch.float32) + 0.5,
            torch.arange(self.w / 2, dtype=torch.float32) + 0.5,
            indexing='ij'
        )
        self.base_pixel_grid = torch.stack([x, y, torch.ones_like(x)], dim=-1)

    def __len__(self):
        return len(self.segments)

    def _get_raymap(self, K, c2w):
        """
        Optimized Raymap generation using pre-computed grid.
        """
        R_c2w = torch.from_numpy(c2w[:3, :3]).float()
        t_c2w = torch.from_numpy(c2w[:3, 3]).float()
        K_inv = torch.from_numpy(np.linalg.inv(K)).float()

        # (H, W, 3) @ (3, 3).T -> (H, W, 3)
        rays_opt = self.base_pixel_grid @ K_inv.T
        rays_world = rays_opt @ R_c2w.T
        origins_world = t_c2w.view(1, 1, 3).expand(self.h, self.w, 3)

        return torch.cat([rays_world, origins_world], dim=-1)

    def _get_sequenced_raymap(self, K, seq_extrinsic):
        R_w2c = seq_extrinsic[:, :3, :3]
        t_c2w = -R_w2c.mT @ seq_extrinsic[:, :3, 3:4]
        K_inv = K.inverse()
        rays_opt = self.base_pixel_grid @ K_inv.mT[:, None]
        rays_world = rays_opt @ R_w2c[:, None].float()
        origins_world = t_c2w.view(-1, 1, 1, 3).expand(-1, self.h // 2, self.w // 2, 3)
        return torch.cat([rays_world, origins_world], dim=-1)

    @staticmethod
    def _load_pickle(path):
        with open(path, 'rb') as f:
            return pickle.load(f)

    @staticmethod
    def _load_sparse_depth(path):
        sparse_matrix = scipy.sparse.load_npz(path)
        depth_dense = sparse_matrix.toarray()
        return depth_dense

    @staticmethod
    def _view_mat_to_normed_extrinsic(c2w: torch.Tensor) -> tuple[torch.Tensor, float]:
        ex_t_norm = c2w.inverse() @ c2w[:1]
        c2ws = ex_t_norm.inverse()
        translations = c2ws[:, :3, 3]
        dists = translations.norm(dim=-1)
        median_dist = torch.median(dists)
        median_dist = torch.clamp(median_dist, min=1e-1)
        ex_t_norm[:, :3, 3] = ex_t_norm[:, :3, 3] / median_dist
        return ex_t_norm, median_dist

    def __getitem__(self, idx):
        segment_path = self.segments[idx]

        # --- Load Metadata ---
        v2w_list = self._load_pickle(os.path.join(segment_path, 'v2w.pkl'))
        cameras_data = self._load_pickle(os.path.join(segment_path, 'cameras.pkl'))
        times_data = self._load_pickle(os.path.join(segment_path, 'times.pkl'))

        total_available_frames = len(times_data)

        images_seq, rays_seq, depths_seq, c2w_seq, c2v_seq, intrinsics, cam_id_seq, timestamp, rgbs, minidepths = [], [], [], [], [], [], [], [], [], []
        cam2v_dict = cameras_data['cam2v']

        start_idx = np.random.randint(0, total_available_frames - self.seq_len * self.step)

        for source_idx in range(start_idx, start_idx + self.seq_len * self.step, self.step):
            v2w_curr = v2w_list[source_idx]

            # Iterate Cameras
            for c, cam_name in enumerate(self.camera_names):
                # Pose
                c2v = cam2v_dict[cam_name] @ T_waymo2colmap  # cam2vehicle in colmap convention
                c2v_seq.append(c2v)
                c2w = np.matmul(v2w_curr, c2v)
                c2w_seq.append(c2w)

                # Image & Depth Loading
                file_id = str(source_idx).zfill(6)

                # Image
                img_path = os.path.join(segment_path, cam_name, f"{file_id}.jpg")
                image = Image.open(img_path).convert('RGB')
                image = np.array(image)

                # Depth
                depth_path = os.path.join(segment_path, cam_name, f"{file_id}_z.npz")
                depth = self._load_sparse_depth(depth_path)

                rgb_out, depth_out, depth_out_x2, K = process_rgb_and_depth(image, depth, cameras_data['K'][cam_name],
                                                                            cameras_data['distortion'][cam_name], self.target_size, crop_to_valid=False)  # TODO: crop_to_valid true for side cams

                if self.debug:
                    rgbs.append(rgb_out)
                # ray_maps = self._get_raymap(K, c2w)  # TODO: ray maps should be calculated on normalized extrinsics!
                images_seq.append(self.transform(rgb_out))
                # rays_seq.append(ray_maps)
                depths_seq.append(depth_out)
                minidepths.append(depth_out_x2)
                intrinsics.append(K)

                cam_id_seq.append(c)
                timestamp.append(times_data[source_idx])

        # normalize timestamps
        timestamp = np.array(timestamp)
        timestamp -= timestamp[0]

        c2w = torch.from_numpy(np.stack(c2w_seq)).float()
        c2v = torch.from_numpy(np.stack(c2v_seq)).float()
        normed_extrinsic, scale = self._view_mat_to_normed_extrinsic(c2w)  # w2c
        intrinsics = torch.from_numpy(np.stack(intrinsics)).float()

        rays = self._get_sequenced_raymap(intrinsics, normed_extrinsic)
        data = {
                "images": torch.stack(images_seq),  # (N * 5, 3, H, W)
                "ray": rays,   # (N * 5, H, W, 6)
                "depth": torch.stack(depths_seq).float(),  # (N * 5, H, W)
                "mini_depth": torch.stack(minidepths).float(),  # (N * 5, H, W)
                "normed_extrinsic": normed_extrinsic,  # (N * 5, 4, 4)
                "rig_extrinsic": c2v.inverse(),
                "scale": scale,  # (1, )
                "intrinsics": intrinsics,  # (N * 5, 3, 3)
                "camera_id": torch.tensor(cam_id_seq),  # N * 5
                "timestamps": torch.from_numpy(timestamp).float()  # N * 5
            }
        if self.debug:
            data['rgb'] = np.stack(rgbs)
            data['c2w'] = c2w

        return data

```

`process_rgb_and_depth.py` file

```python
def process_rgb_and_depth(img_rgb, img_depth, K, D, target_shape=None, crop_to_valid=True):
    """
    Undistorts and resizes RGB and DEPTH simultaneously.

    Args:
        img_rgb: (H, W, 3) uint8
        img_depth: (H, W) float32 (or similar)
        K: Intrinsics (3,3)
        D: Distortion coeffs
        target_shape: (H, W) target size (divisible by 14)
        crop_to_valid: whether to crop black pixels

    Returns:
        rgb_out, depth_out, depth__out_x2, K_new
    """
    h, w = img_rgb.shape[:2]

    # --- 1. Compute the Mapping ---
    # We get the new optimal camera matrix and the ROI crop
    new_K = K.copy()
    new_K[0, 2] = w / 2.0  # Force cx to center
    new_K[1, 2] = h / 2.0  # Force cy to center

    # initUndistortRectifyMap calculates the "lookup table" for where every pixel
    # in the new image comes from in the old image.
    map_x, map_y = cv2.initUndistortRectifyMap(K, D, None, new_K, (w, h), cv2.CV_32FC1)

    # --- 2. Apply Mapping (Undistort) ---
    # RGB uses LINEAR (smooths colors)
    rgb_undist = cv2.remap(img_rgb, map_x, map_y, interpolation=cv2.INTER_AREA)

    # DEPTH uses NEAREST (preserves exact values, avoids edge artifacts)
    # Note: If your depth is dense and smooth, LINEAR is acceptable,
    # but NEAREST is safer for geometric consistency.
    depth_undist = cv2.remap(img_depth, map_x, map_y, interpolation=cv2.INTER_NEAREST)

    # --- 3. Crop (If alpha=0) ---
    if crop_to_valid:
        dy = 200
        rgb_undist = rgb_undist[dy:]
        depth_undist = depth_undist[dy:]

        # Adjust Principal Point for the crop
        new_K[1, 2] -= dy
        h -= dy

    # --- 4. Determine Target Size (Divisible by 14) ---
    if target_shape is None:
        new_h = int(round(h / 14) * 14)
        new_w = int(round(w / 14) * 14)
    else:
        new_h, new_w = target_shape

    # --- 5. Resize ---
    # Resize RGB (Linear)
    rgb_out = cv2.resize(rgb_undist, (new_w, new_h), interpolation=cv2.INTER_AREA)

    # Resize Depth (Nearest)
    depth_out = resize_z_depth(torch.from_numpy(depth_undist)[None], target_size=(new_h, new_w))[0]
    depth_out_x2 = resize_z_depth(torch.from_numpy(depth_undist)[None], target_size=(new_h//2, new_w//2))[0]

    # --- 6. Scale Intrinsics ---
    scale_x = new_w / w
    scale_y = new_h / h

    K_final = new_K.copy()
    K_final[0, 0] *= scale_x  # fx
    K_final[1, 1] *= scale_y  # fy
    K_final[0, 2] *= scale_x  # cx
    K_final[1, 2] *= scale_y  # cy

    return rgb_out, depth_out, depth_out_x2, K_final


def resize_z_depth(depth_tensor, target_size=(336, 504)):
    """
    Correctly downsamples sparse Z-Depth by keeping the CLOSEST valid point.
    depth_tensor: [B, 1, H, W] (Positive Z values, 0=invalid)
    """
    # 1. Replace 0 (invalid) with -Infinity
    #    We do this so they are ignored during the "max" calculation.
    #    Valid depths (e.g. 5.0) become negative (-5.0).
    #    -5.0 is greater than -50.0, so "max" will pick the closer object.
    depth_inverted = torch.where(depth_tensor > 0, -depth_tensor, torch.tensor(float('-inf')))

    # 2. Max Pool (which effectively selects the 'least negative' aka closest value)
    pooled_inverted = F.adaptive_max_pool2d(depth_inverted, target_size)

    # 3. Flip back to positive
    pooled_depth = -pooled_inverted

    # 4. Clean up: Convert Infinity back to 0 (empty pixels)
    pooled_depth[pooled_depth == float('inf')] = 0.0

    return pooled_depth

```

**Comment by ckLibra** [comment]
Thanks for your sharing. But I found that when using the get_raymap function to convert the extrinsic and intrinsic to ray map, and then using the get_extrinsic_from_camray function in the project to get the estimated extrinsic and intrinsic, the estimated extrinsics and intrinsics could not aligned with the original one, did you have the same problem? @Golbstein

**Comment by anhquancao** [comment]
Hi! Not one of the authors, but in case it helps anyone here. We implemented the full training pipeline and losses for DA3 following the paper details. Hopefully is is a useful starting point:

https://github.com/valeoai/OccAny#occany-depth-anything-3--sam3-1

More context in https://github.com/ByteDance-Seed/Depth-Anything-3/issues/35#issuecomment-4825419485
